job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,256 @@
+      // https://github.com/spark-jobserver/spark-jobserver/issues/349
+
+      startContext(contextName, mergedConfig, true) { ref =>
+        originator ! contexts(contextName)
+      } { err =>
+        originator ! ContextInitError(err)
+      }
+
+    case GetResultActor(name) =>
+      sender ! contexts.get(name).map(_._2).getOrElse(globalResultActor)
+
+    case GetContext(name) =>
+      if (contexts contains name) {
+        sender ! contexts(name)
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case StopContext(name) =>
+      if (contexts contains name) {
+        logger.info("Shutting down context {}", name)
+        contexts(name)._1 ! PoisonPill
+        sender ! ContextStopped
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case Terminated(actorRef) =>
+      val name: String = actorRef.path.name
+      logger.info("Actor terminated: {}", name)
+      contexts.retain { case (name, (jobMgr, resActor)) => jobMgr != actorRef }
+  }
+
+  private def initContext(actorName: String, ref: ActorRef, timeoutSecs: Long = 1)
+                         (isAdHoc: Boolean,
+                          successFunc: ActorRef => Unit,
+                          failureFunc: Throwable => Unit): Unit = {
+    import akka.pattern.ask
+
+    val resultActor = if (isAdHoc) globalResultActor else context.actorOf(Props(classOf[JobResultActor]))
+    (ref ? JobManagerActor.Initialize(
+      daoActor, Some(resultActor)))(Timeout(timeoutSecs.second)).onComplete {
+      case Failure(e:Exception) =>
+        logger.info("Failed to send initialize message to context " + ref, e)
+        ref ! PoisonPill
+        failureFunc(e)
+      case Success(JobManagerActor.InitError(t)) =>
+        logger.info("Failed to initialize context " + ref, t)
+        ref ! PoisonPill
+        failureFunc(t)
+      case Success(JobManagerActor.Initialized(ctxName, resActor)) =>
+        logger.info("SparkContext {} joined", ctxName)
+        contexts(ctxName) = (ref, resActor)
+        context.watch(ref)
+        successFunc(ref)
+    }
+  }
+
+  private def startContext(name: String, contextConfig: Config, isAdHoc: Boolean)
+                          (successFunc: ActorRef => Unit)(failureFunc: Throwable => Unit) = {
+    require(!(contexts contains name), "There is already a context named " + name)
+    val contextActorName = "jobManager-" + java.util.UUID.randomUUID().toString.substring(16)
+
+    logger.info("Starting context with actor name {}", contextActorName)
+
+    val contextDir = createContextDir(name, contextConfig, isAdHoc, contextActorName)
+    val pb = Process(s"$managerStartCommand $contextDir ${selfAddress.toString}")
+    val pio = new ProcessIO(_ => (),
+                        stdout => scala.io.Source.fromInputStream(stdout)
+                          .getLines.foreach(println),
+                        stderr => scala.io.Source.fromInputStream(stderr).getLines().foreach(println))
```

Comment:
logger instead of println?

```
val pio = new ProcessIO(_ => (),
                        stdout => scala.io.Source.fromInputStream(stdout)
                          .getLines.foreach(logger.info(_)),
                        stderr => scala.io.Source.fromInputStream(stderr).getLines().foreach(logger.error(_)))
```